services/worker-manager/src/providers/aws.js

```diff
@@ -82,83 +91,93 @@ class AwsProvider extends Provider {
       });
     }
 
-    // Make sure we don't get "The same resource type may not be specified more than once in tag specifications" errors
-    const TagSpecifications = config.launchConfig.TagSpecifications ? config.launchConfig.TagSpecifications : [];
-    const instanceTags = [];
-    const otherTagSpecs = [];
-    TagSpecifications.forEach(ts =>
-      ts.ResourceType === 'instance' ? instanceTags.concat(ts.Tags) : otherTagSpecs.push(ts)
-    );
+    const shuffledConfigs = _.shuffle(workerPool.config.launchConfigs);
 
     let spawned;
-    try {
-      spawned = await ec2.runInstances({
-        ...config.launchConfig,
-
-        UserData: userData.toString('base64'), // The string needs to be base64-encoded. See the docs above
-
-        MaxCount: toSpawn,
-        MinCount: toSpawn,
-        TagSpecifications: [
-          ...otherTagSpecs,
-          {
-            ResourceType: 'instance',
-            Tags: [
-              ...instanceTags,
-              {
-                Key: 'CreatedBy',
-                Value: `taskcluster-wm-${this.providerId}`,
-              }, {
-                Key: 'Owner',
-                Value: workerPool.owner,
-              },
-              {
-                Key: 'ManagedBy',
-                Value: 'taskcluster',
-              },
-              {
-                Key: 'Name',
-                Value: `${workerPoolId}`,
-              }],
-          },
-        ],
-      }).promise();
-    } catch (e) {
-      this.monitor.err(`Error calling AWS API: ${e}`);
+    let toSpawnCounter = toSpawn;
+    for await (let config of shuffledConfigs) {
+      // Make sure we don't get "The same resource type may not be specified
+      // more than once in tag specifications" errors
+      const TagSpecifications = config.TagSpecifications ? config.TagSpecifications : [];
+      const instanceTags = [];
+      const otherTagSpecs = [];
+      TagSpecifications.forEach(ts =>
+        ts.ResourceType === 'instance' ? instanceTags.concat(ts.Tags) : otherTagSpecs.push(ts)
+      );
+
+      try {
+        spawned = await this.ec2s[config.region].runInstances({
+          ...config.launchConfig,
+
+          UserData: userData.toString('base64'), // The string needs to be base64-encoded. See the docs above
+
+          MaxCount: Math.min(toSpawnCounter, toSpawnPerConfig),
```

Comment:
Can `toSpawnCounter` end up being negative here I think? Probably we can get away with just passing `toSpawnPerConfig` directly to `MaxCount` and exiting this loop early if `toSpawnCounter <= 0`?

Reply:
So the states where `toSpawnCounter <= 0` are reached when unforeseen happens, like instances shutting themselves down immediately etc. I'm not noticing that affecting the functionality, but it does create error noise. I'll add a check at the beginning of the loop and will be breaking out of the loop immediately if `toSpawnCounter <= 0`